### PR TITLE
Refactor of Upper Mac

### PR DESCRIPTION
**Pull request by marenz2569**

### CODE REVIEW

File: include/l2/logical_channel.hpp [hunk 1]
```diff
@@ -18,6 +18,21 @@ enum class LogicalChannel {
     kStealingChannel
 };
 
+constexpr auto to_string(LogicalChannel channel) noexcept -> const char* {
```
Comment (tanneberger) [inline include/l2/logical_channel.hpp:21]: Why not turn this into an interface?
```c++
class Serialize {
public:
    virtual auto to_string(LogicalChannel channel) noexcept -> const char* = 0;
}
```

?

File: include/l2/upper_mac.hpp [hunk 2]
```diff
@@ -32,122 +24,28 @@ class UpperMac {
...
-    // hashmap to keep track of framented mac segments
-    std::unordered_map<AddressType, std::vector<BitVector>> fragment_map_ = {};
-    std::vector<BitVector> fragment_list_{};
-    bool fragment_end_received_{};
-    bool fragment_end_hu_received_{};
-    AddressType last_address_type_{};
-    // save the last MAC-ACCESS or MAC-DATA where reservation_requirement is 0b0000 (1 sublot) for END-HU
-    AddressType last_address_type_end_hu_{};
-};
+    std::shared_ptr<Reporter> reporter_;
```
Comment (tanneberger) [inline include/l2/upper_mac.hpp:46]: why shared?

File: include/l2/upper_mac_packet_builder.hpp [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    /// slots.
+    /// \param other the UpperMacPackets which should be merged into the current one
+    auto merge(UpperMacPackets&& other) -> void {
+        std::move(other.c_plane_signalling_packets_.begin(), other.c_plane_signalling_packets_.end(),
+                  std::back_inserter(c_plane_signalling_packets_));
+        std::move(other.u_plane_signalling_packet_.begin(), other.u_plane_signalling_packet_.end(),
+                  std::back_inserter(u_plane_signalling_packet_));
+
+        if (u_plane_traffic_packet_ && other.u_plane_traffic_packet_) {
+            throw std::runtime_error("Trying to merge two packets both with a UpperMacUPlaneTrafficPacket");
```
Comment (tanneberger) [inline include/l2/upper_mac_packet_builder.hpp:40]: I think in the long run we probably shouldn't use hard exceptions. 

Is it possible to create a malformed packet which would take the software down?

File: include/l2/upper_mac_packet_builder.hpp [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        if (other.broadcast_packet_) {
+            broadcast_packet_ = std::move(*other.broadcast_packet_);
+        }
+    }
+
+    /// Check if the packet contains data that is of importance for C-Plane or U-Plane
+    /// \return true if the UpperMacPackets contain user or control plane data (either signalling or traffic)
+    [[nodiscard]] auto has_user_or_control_plane_data() const -> bool {
+        for (const auto& packet : c_plane_signalling_packets_) {
+            if (packet.is_null_pdu()) {
```
Comment (tanneberger) [inline include/l2/upper_mac_packet_builder.hpp:60]: I thing the other way around is more understandable.


```c++
if (!packet.is_null_pdu()){ 
    return true;
}
```

**Review by tanneberger — APPROVED** [review]

File: include/l2/upper_mac_packet_builder.hpp (hunk 3 above)
Comment (marenz2569) [reply include/l2/upper_mac_packet_builder.hpp:40]: As far as I read the standard this should not happen. We will discard Slots that trigger a runtime_error during parsing. The expectation is that these slots will be logged. See https://github.com/tlm-solutions/tetra-decoder/issues/15 and https://github.com/tlm-solutions/tetra-decoder/issues/16

File: include/l2/upper_mac.hpp (hunk 2 above)
Comment (marenz2569) [reply include/l2/upper_mac.hpp:46]: this is due to older code architecture. It needs to be refactored in another PR

File: include/l2/logical_channel.hpp (hunk 1 above)
Comment (marenz2569) [reply include/l2/logical_channel.hpp:21]: as discussed. lets leave it like it is to not lose the constexpr

**Pull request merged**